src/hotspot/share/opto/mempointer.hpp

```diff
@@ -0,0 +1,487 @@
+    _distance(distance)
+  {
+    const jint max_distance = 1 << 30;
+    assert(_distance < max_distance && _distance > -max_distance, "safe distance");
+  }
+
+public:
+  MemPointerAliasing() : MemPointerAliasing(Unknown, 0) {}
+
+  static MemPointerAliasing make_unknown() {
+    return MemPointerAliasing();
+  }
+
+  static MemPointerAliasing make_always(const jint distance) {
+    return MemPointerAliasing(Always, distance);
+  }
+
+  // Use case: exact aliasing and adjacency.
+  bool is_always_at_distance(const jint distance) const {
+    return _aliasing == Always && _distance == distance;
+  }
+
+#ifndef PRODUCT
+  void print_on(outputStream* st) const {
+    switch(_aliasing) {
+      case Unknown: st->print("Unknown");               break;
+      case Always:  st->print("Always(%d)", _distance); break;
+      default: ShouldNotReachHere();
+    }
+  }
+#endif
+};
+
+// Summand of a MemPointerDecomposedForm:
+//
+//   summand = scale * variable
+//
+class MemPointerSummand : public StackObj {
+private:
+  Node* _variable;
+  NoOverflowInt _scale;
+
+public:
+  MemPointerSummand() :
+      _variable(nullptr),
+      _scale(NoOverflowInt::make_NaN()) {}
+  MemPointerSummand(Node* variable, const NoOverflowInt scale) :
+      _variable(variable),
+      _scale(scale)
+  {
+    assert(_variable != nullptr, "must have variable");
+    assert(!_scale.is_zero(), "non-zero scale");
+  }
+
+  Node* variable() const { return _variable; }
+  NoOverflowInt scale() const { return _scale; }
+
+  static int cmp_for_sort(MemPointerSummand* p1, MemPointerSummand* p2) {
+    if (p1->variable() == nullptr) {
+      return (p2->variable() == nullptr) ? 0 : 1;
+    } else if (p2->variable() == nullptr) {
+      return -1;
+    }
+
+    return p1->variable()->_idx - p2->variable()->_idx;
+  }
+
+  friend bool operator==(const MemPointerSummand a, const MemPointerSummand b) {
+    // Both "null" -> equal.
+    if (a.variable() == nullptr && b.variable() == nullptr) { return true; }
+
+    // Same variable and scale?
+    if (a.variable() != b.variable()) { return false; }
+    return a.scale() == b.scale();
+  }
+
+  friend bool operator!=(const MemPointerSummand a, const MemPointerSummand b) {
+    return !(a == b);
+  }
+
+#ifndef PRODUCT
+  void print_on(outputStream* st) const {
+    st->print("Summand[");
+    _scale.print_on(st);
+    tty->print(" * [%d %s]]", _variable->_idx, _variable->Name());
+  }
+#endif
+};
+
+// Decomposed form of the pointer sub-expression of "pointer".
+//
+//   pointer = con + sum(summands)
+//
+class MemPointerDecomposedForm : public StackObj {
+private:
+  // We limit the number of summands to 10. Usually, a pointer contains a base pointer
+  // (e.g. array pointer or null for native memory) and a few variables. For example:
+  //
+  //   array[j]                      ->  array_base + j + con              -> 2 summands
+  //   nativeMemorySegment.get(j)    ->  null + address + offset + j + con -> 3 summands
```

Comment:
Does this means it supports only Unsafe and Array access?

Reply:
I removed these examples, but instead added extensive examples at the beginning of `mempointer.hpp`